File: csharp/0494目标和.cs

public partial class Solution
{
    public int FindTargetSumWays(int[] nums, int target)
    {
        int res = 0;
        int size = nums.Length;
        void _backTracking1(int start=0,int sum=0)
        {
            if(start==size)
            {
                if(sum == target)
                    res++;
                return;
            }

            _backTracking1(start + 1, sum + nums[start]);
            _backTracking1(start + 1, sum - nums[start]);
        }
        
        int total = nums.Sum();

        void _backTracking2(int start,int sum)
        {
            if(sum==target)
            {
                res++;
            }    

            for (int i = start; i < size; i++)
            {
                if(sum<target)
                    return;
                _backTracking2(i + 1, sum - 2 * nums[i]);
            }
        }

        _backTracking2(0,total);

        return res;
    }


}

File: python/main/0494目标和.py

def findTargetSumWays(nums: list, target: int) -> int:
    total=sum(nums)
    
    # neg非负
    if(total-target<0):
        return 0
    # neg是整数
    if((total-target)%2):
        return 0

    # total-neg-neg=target
    neg=(total-target)//2

    # 转化为从nums中凑出等于neg的情况总数
    # 01背包组合
    m=len(nums)
    n=neg+1

    # dp=[1]+[0 if nums[0]!=i else 1 for i in range(1,n)]
    dp=[1]+[0]*neg
    
    for i in range(0,m):
        for j in range(n-1,nums[i]-1,-1):
            dp[j]+=dp[j-nums[i]]
        
    return dp[-1]

print(findTargetSumWays([0,0,0,0,0,0,0,0,1],1))
